model_compression_toolkit/gptq/common/gptq_training.py

```diff
@@ -132,25 +144,36 @@ def compute_hessian_based_weights(self,
+            existing_computations_count = hessian_service.count_cache(self.gptq_hessian_config)
+            images_to_compute_count = number_of_images - existing_computations_count
 
             points_apprx_jacobians_weights = []
-            for i in range(1, images.shape[0] + 1):
-                Logger.info(f"Computing Jacobian-based weights approximation for image sample {i} out of {images.shape[0]}...")
-                # Note that in GPTQ loss weights computation we assume that there aren't replacement output nodes,
-                # therefore, output_list is just the graph outputs, and we don't need the tuning factor for
-                # defining the output weights (since the output layer is not a compare point).
-                image_ip_gradients = self.fw_impl.model_grad(self.graph_float,
-                                                             {inode: self.fw_impl.to_tensor(images[i - 1:i]) for inode
-                                                              in
-                                                              self.graph_float.get_inputs()},
-                                                             self.compare_points,
-                                                             output_list=model_output_replacement,
-                                                             all_outputs_indices=[],
-                                                             alpha=0,
-                                                             norm_weights=self.gptq_config.hessian_weights_config.norm_weights,
-                                                             n_iter=self.gptq_config.hessian_weights_config.hessians_n_iter)
+
+            # Use existing computations for the config to save time
+            existing_computations = hessian_service.fetch_hessian(self.gptq_hessian_config)
+            for image_id, node_to_score in existing_computations.items():
+                image_ip_gradients = []
+                for ip in self.compare_points:
+                    image_ip_gradients.append(node_to_score[ip])
                 points_apprx_jacobians_weights.append(image_ip_gradients)
+
+            if images_to_compute_count > 0:
+                newly_computed = 0
+                for i in range(1, images.shape[0] + 1):
+                    Logger.info(f"Computing Jacobian-based weights approximation for image sample {i} out of {images.shape[0]}...")
+                    images = [images[i - 1:i]] * len(self.graph_float.get_inputs())  # TODO: change to image per input
+                    hessian_data = hessian_service.fetch_hessian(self.gptq_hessian_config,
+                                                                 images[i - 1:i])
+                    image_ip_gradients = []
+                    for ip in self.compare_points:
+                        image_ip_gradients.append(hessian_data[ip])
+                    points_apprx_jacobians_weights.append(image_ip_gradients)
```

Comment:
Same regarding "jacobians" (change to "hessian").
Also, regarding the term "weights", let's decide on a different term and change it here as well.